File: src/main/java/ufcg/controllers/Facade.java
package ufcg.controllers;

import java.util.List;
import java.util.Scanner;

import ufcg.entities.Aluno;
import ufcg.entities.Grupo;
import ufcg.repositories.AlunoRepository;
import ufcg.repositories.GrupoRepository;
import ufcg.repositories.RespondenteRepository;

public class Facade {
    GrupoRepository repositorioGrupos;
    AlunoRepository repositorioAlunos;
    RespondenteRepository repositorioRespondentes;

    public Facade() {
        this.repositorioAlunos = new AlunoRepository();
        this.repositorioGrupos = new GrupoRepository();
        this.repositorioRespondentes = new RespondenteRepository();
    }

    public void cadastrarAluno(Scanner sc) throws Exception {
        System.out.print("\nMatrícula: ");
        String matricula = sc.next();

        System.out.print("\nNome: ");
        String nome = sc.next();

        System.out.print("\nCurso: ");
        String curso = sc.next();

        repositorioAlunos.cadastrarAluno(nome, matricula, curso);
        System.out.print("\nCadastro realizado!");
    }

    public void consultarAluno(Scanner sc) {
        System.out.print("\nMatrícula: ");
        String matricula = sc.next();

        Aluno alunoEncontrado = repositorioAlunos.getAluno(matricula);
        if(alunoEncontrado == null) System.out.print("\nAluno não cadastrado");
        else System.out.print("\n" + alunoEncontrado);
    }

    public void cadastrarGrupo(Scanner sc) throws Exception {
        System.out.print("\nGrupo: ");
        String nome = sc.next();
        System.out.print("\nTamanho: ");
        int tamanho = sc.nextInt();

        repositorioGrupos.cadastrarGrupo(nome, tamanho);
    }

    public void alocarAlunoEmGrupos(Scanner sc) throws Exception {
        System.out.print("\n(A)locar Aluno ou (P)ertinênncia a Grupo?");
        String subOpcao = sc.next();

        if(subOpcao.toLowerCase().equals("a")) {
            System.out.print("\nMatrícula: ");
            String matricula = sc.next();

            Aluno alunoEncontrado = repositorioAlunos.getAluno(matricula);
            if(alunoEncontrado == null) System.out.print("\n Aluno não cadastrado");

            System.out.print("\nGrupo: ");
            String nomeGrupo = sc.next();

            repositorioGrupos.alocarAlunoEmGrupo(alunoEncontrado, nomeGrupo);
            System.out.print("\nAluno alocado");
        } else if(subOpcao.toLowerCase().equals("p")) {
            System.out.print("\nGrupo: ");
            String nomeGrupo = sc.next();

            System.out.print("\nAluno: ");
            String matriculaAluno = sc.next();

            Grupo grupoEncontrado = repositorioGrupos.getGrupo(nomeGrupo);
            if(grupoEncontrado == null) System.out.print("\nGrupo não cadastrado");

            Aluno alunoEncontrado = repositorioAlunos.getAluno(matriculaAluno);
            if(alunoEncontrado == null) System.out.print("\nAluno não registrado");
            
            if(grupoEncontrado.isAlunoEmGrupo(alunoEncontrado)) System.out.print("\nAluno pertence ao grupo");
            else System.out.print("\nAluno não pertence ao grupo");
        }
    }

    public void cadastrarAlunosQueRespondemQuestoesNoQuadro(Scanner sc) {
        System.out.print("\nMatricula: ");
        String matricula = sc.next();

        Aluno alunoEncontrado = repositorioAlunos.getAluno(matricula);
        if(alunoEncontrado == null) System.out.print("\nAluno não existe");
        else repositorioRespondentes.registrarRespondente(alunoEncontrado);
        System.out.print("\nAluno registrado");
    }

    public void imprimirAlunosQueRespondemQuestoesNoQuadro(Scanner sc) {
        List<Aluno> respondentes = repositorioRespondentes.getRegistroRespondentes();
        for(int i = 0; i < respondentes.size(); i++) {
            System.out.print("\n" + (i+1) + ". " + respondentes.get(i));
        }
    }

    public void checagemDosGruposDosAlunos(Scanner sc) throws Exception {
        System.out.print("\nAluno: ");
        String matricula = sc.next();

        Aluno alunoEncontrado = repositorioAlunos.getAluno(matricula);
        if(alunoEncontrado == null) throw new Exception("Aluno não existe");
        List<Grupo> gruposDoQualFazParte = repositorioGrupos.getGruposDoQualFazParte(alunoEncontrado);

        System.out.print("\nGrupos: ");
        for(Grupo grupo : gruposDoQualFazParte) System.out.print("\n- " + grupo.getNome());
    }

    public void imprimirEstatisticas(Scanner sc) {
        List<Aluno> todosOsAlunosCadastrados = repositorioAlunos.getTodosAlunosCadastrados();
        String relatorio = repositorioRespondentes.getRelatorioEstatisticas(todosOsAlunosCadastrados);
        System.out.print("\n"+relatorio);
    }
}

File: rust/src/facade.rs
use crate::utils::{
    io,
    parsers
};
use crate::repositories::{
    alunos::RepositorioAlunos,
    grupos::RepositorioGrupos,
    respondentes::RepositorioRespondentes
};
use crate::models::{
    aluno::Aluno,
};
pub struct Facade {
    repositorio_alunos: RepositorioAlunos,
    repositorio_grupos: RepositorioGrupos,
    repositorio_respondentes: RepositorioRespondentes
}

impl Facade {
    pub fn new() -> Self {
        Facade {
            repositorio_alunos: RepositorioAlunos::new(),
            repositorio_grupos: RepositorioGrupos::new(),
            repositorio_respondentes: RepositorioRespondentes::new()
        }
    }

    pub fn cadastrar_aluno(&mut self) {
        print!("Matrícula: "); io::flush_stdout_buffer();
        let matricula: String = io::input();
    
        print!("Nome: "); io::flush_stdout_buffer();
        let nome: String = io::input();
    
        print!("Curso: "); io::flush_stdout_buffer();
        let curso: String = io::input();
    
        // might cast an error:
        match &self.repositorio_alunos.cadastrar_aluno(matricula, nome, curso) {
            Ok(_) => println!("Cadastro Realizado!"),
            Err(_error) => println!("Matrícula já Cadastrada!"),
        }
    }

    pub fn consultar_aluno(&self) {
        print!("Matrícula: "); io::flush_stdout_buffer();
        let matricula:String = io::input();
        println!();
        match &self.repositorio_alunos.procurar_aluno(&matricula) {
            Some(aluno) => aluno.display(),
            None => println!("Aluno não cadastrado")
        }
    }

    pub fn cadastrar_grupo(&mut self) {
        print!("Grupo: "); io::flush_stdout_buffer();
        let nome_grupo: String = io::input();
        print!("Tamanho: "); io::flush_stdout_buffer();
        let tamanho: isize = parsers::parse_to_int(&io::input());

        self
            .repositorio_grupos
            .cadastrar_novo_grupo(nome_grupo, tamanho);

        println!("Cadastro Realizado");
    }

    pub fn alocar_aluno_em_grupo(&mut self) {
        print!("(A)locar Aluno ou (P)ertinência a Grupo? "); io::flush_stdout_buffer();
        let user_option: String = io::input().to_lowercase();

        match user_option.as_str().trim() {
            "a" => {
                print!("Matrícula: "); io::flush_stdout_buffer();
                let matricula: String = io::input();
                print!("Grupo: "); io::flush_stdout_buffer();
                let nome_grupo: String = io::input();

                let aluno: Option<&Aluno> = self.repositorio_alunos.procurar_aluno(&matricula);
                match aluno {
                    Some(unwraped_aluno) => {
                        self.repositorio_grupos.alocar_aluno_em_grupo(unwraped_aluno.clone(), &nome_grupo);
                    },
                    None => println!("Aluno não cadastrado!"),
                };
            },
            "p" => {
                print!("Grupo: "); io::flush_stdout_buffer();
                let nome_grupo: String = io::input();
                print!("Aluno: "); io::flush_stdout_buffer();
                let matricula_aluno: String = io::input();

                if !self.repositorio_alunos.aluno_already_registered(&matricula_aluno) {
                    println!("Aluno não existe");
                    return;
                }

                if self.repositorio_grupos.verificar_pertinencia_a_grupo(&nome_grupo, &matricula_aluno) {
                    println!("Aluno pertence ao grupo");
                } else {
                    println!("Aluno não pertence ao grupo");
                }
            }
            _ => println!("Opção inválida"),
        }
    }

    pub fn cadastrar_respondente(&mut self) {
        print!("Matricula: "); io::flush_stdout_buffer();
        let matricula: String = io::input();

        match &self.repositorio_respondentes.adicionar_respondente(matricula, &self.repositorio_alunos) {
            Ok(_) => println!("Aluno Registrado!"),
            Err(_) => println!("Matricula informada não existe no sistema")
        }
    }
}